pubsub/azuresb/azuresb.go

```diff
@@ -506,16 +524,44 @@ func (s *subscription) updateMessageDispositions(ctx context.Context, ids []driv
 		return nil
 	}
 
-	lockIds := []amqp.UUID{}
-	for _, mid := range ids {
-		if id, ok := mid.(*uuid.UUID); ok {
-			lockTokenBytes := [16]byte(*id)
-			lockIds = append(lockIds, amqp.UUID(lockTokenBytes))
+	// Group by partitionID; AzureSB doesn't support updating dispositions for
+	// messages from different partitions.
+	partitions := map[int16][]amqp.UUID{}
+	for _, ackID := range ids {
+		if pid, ok := ackID.(*partitionAckID); ok {
+			lockTokenBytes := [16]byte(*pid.LockToken)
+			partitions[pid.PartitionID] = append(partitions[pid.PartitionID], amqp.UUID(lockTokenBytes))
 		}
 	}
+
+	// Update partitions in parallel.
+	const maxConcurrency = 5
+	sem := make(chan struct{}, maxConcurrency)
+	var mu sync.Mutex
+	var lastErr error
+	for _, lockTokens := range partitions {
+		sem <- struct{}{}
+		go func() {
+			defer func() { <-sem }() // Release the semaphore.
+			if err := s.updateMessageDispositionsInPartition(ctx, lockTokens, disposition); err != nil {
+				mu.Lock()
+				defer mu.Unlock()
+				lastErr = err
```

Comment:
Would it be possible to maintain the errors if multiple occur?  

Combine the strings maybe?

My only concern is multiple messages acks fail, and not seeing which ones/or their errors in the logs may be an issue.

Reply:
Sure; done. Although:

* (nit) It's not multiple acks, it's multiple batches of acks. So even if you get an error, it's not obvious how many acks actually failed (some of the batch may have succeeded). In our benchmark, we regularly see batches of > 100 acks going out at once.

* You'll get this error back from `Receive` or `Shutdown`, not from `Message.Ack`.

Reply:
Thank you.

Correct, I misstated that the failure would be from multiple batchs of ack messages.

Yes, since it happens async the next call to `Receive` (in our use case from #2362 ) we would see this error message.  I do like having all of the errors that occurred since the service bus usually has detailed error strings.